### PR TITLE
Implement powered minecart // minecart furnace

### PR DESCRIPTION
**Pull request by Thorinwasher**
# Description
We all love this entity, that's why I implemented it. As it's definitely one of the most used entities, I have implemented it (this is called irony). Generally furnace minecarts could be considered as the most buggy entity interface in the paper api (If someone is interested, I can show you how bad it is).

Also added some methods that gave compilation errors in PlayerMock  

# Checklist
The following items should be checked before the pull request can be merged.

- [x] Unit tests added.
- [x] Code follows existing code format.

# Info on creating a pull request
- Make sure that unit tests are added which test the relevant changes.
- Make sure that the changes follow the existing code format.

# For maintainer
When a PR is approved, the maintainer should label this PR with `release/*`.

- If the PR fixes a bug in MockBukkit, add the `release/patch` Tag.
- If the PR adds a new feature to MockBukkit, add  the `release/minor` Tag.

Note that a PR that fixes an `UnimplementedOperationException` should be considered a new version and not a bugfix.

### CODE REVIEW

File: src/main/java/be/seeseemelk/mockbukkit/entity/PlayerMock.java [hunk 1]
```diff
@@ -2780,4 +2781,18 @@ public void sendMessage(@NotNull ChatMessageType position, @NotNull BaseComponen
...
+		throw new UnimplementedOperationException();
+	}
+
+	@Override
+	public @NotNull TriState hasFlyingFallDamage()
+	{
+		// TODO Auto-generated method stub
+		throw new UnimplementedOperationException();
+	}
+
```
Comment (thelooter) [inline src/main/java/be/seeseemelk/mockbukkit/entity/PlayerMock.java]: Can these be easily implemented?

File: src/test/java/be/seeseemelk/mockbukkit/entity/PoweredMinecartMockTest.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import be.seeseemelk.mockbukkit.ServerMock;
+
+class PoweredMinecartMockTest
+{
+
+	private ServerMock server;
+	private PoweredMinecartMock minecartFurnace;
+
+	@BeforeEach
+	public void setup()
```
Comment (thelooter) [inline src/test/java/be/seeseemelk/mockbukkit/entity/PoweredMinecartMockTest.java]:
```suggestion
	void setup()
```

File: src/test/java/be/seeseemelk/mockbukkit/entity/PoweredMinecartMockTest.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	{
+		server = MockBukkit.mock();
+		minecartFurnace = new PoweredMinecartMock(server, UUID.randomUUID());
+	}
+
+	@AfterEach
+	public void tearDown()
```
Comment (thelooter) [inline src/test/java/be/seeseemelk/mockbukkit/entity/PoweredMinecartMockTest.java]:
```suggestion
	void tearDown()
```

File: src/test/java/be/seeseemelk/mockbukkit/entity/PoweredMinecartMockTest.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	{
+		MockBukkit.unmock();
+	}
+
+	@Test
+	public void defaultValues()
```
Comment (thelooter) [inline src/test/java/be/seeseemelk/mockbukkit/entity/PoweredMinecartMockTest.java]:
```suggestion
	void defaultValues()
```

File: src/test/java/be/seeseemelk/mockbukkit/entity/PoweredMinecartMockTest.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	{
+		assertEquals(0, minecartFurnace.getFuel());
+		assertEquals(0, minecartFurnace.getPushX());
+		assertEquals(0, minecartFurnace.getPushZ());
+	}
+
+	@Test
+	public void getFuel()
```
Comment (thelooter) [inline src/test/java/be/seeseemelk/mockbukkit/entity/PoweredMinecartMockTest.java]:
```suggestion
	void getFuel()
```

File: src/test/java/be/seeseemelk/mockbukkit/entity/PoweredMinecartMockTest.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+	{
+		minecartFurnace.setFuel(20);
+		assertEquals(20, minecartFurnace.getFuel());
+	}
+
+	@Test
+	public void setFuel_Negative()
```
Comment (thelooter) [inline src/test/java/be/seeseemelk/mockbukkit/entity/PoweredMinecartMockTest.java]:
```suggestion
	void setFuel_Negative()
```

File: src/test/java/be/seeseemelk/mockbukkit/entity/PoweredMinecartMockTest.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+	{
+		assertThrows(IllegalArgumentException.class, () -> minecartFurnace.setFuel(-1));
+	}
+
+	@Test
+	public void getPushX()
```
Comment (thelooter) [inline src/test/java/be/seeseemelk/mockbukkit/entity/PoweredMinecartMockTest.java]:
```suggestion
	void getPushX()
```

File: src/test/java/be/seeseemelk/mockbukkit/entity/PoweredMinecartMockTest.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+	{
+		minecartFurnace.setPushX(1);
+		assertEquals(1, minecartFurnace.getPushX());
+	}
+
+	@Test
+	public void getPushZ()
```
Comment (thelooter) [inline src/test/java/be/seeseemelk/mockbukkit/entity/PoweredMinecartMockTest.java]:
```suggestion
	void getPushZ()
```

**Review by thelooter — CHANGES_REQUESTED** [review]

File: src/main/java/be/seeseemelk/mockbukkit/entity/PlayerMock.java (hunk 1 above)
Comment (Thorinwasher) [reply src/main/java/be/seeseemelk/mockbukkit/entity/PlayerMock.java]: I don't know, maybe. But isn't that part of another pull request? 

(I can implement them if it's fine in this one)

**Comment by Insprill** [comment]
If you go to the files changed view, you can select multiple suggestions to apply with one commit

File: src/main/java/be/seeseemelk/mockbukkit/entity/PlayerMock.java [hunk 9]
```diff
@@ -2780,4 +2783,17 @@ public void sendMessage(@NotNull ChatMessageType position, @NotNull BaseComponen
...
+	{
+		Preconditions.checkNotNull(flyingFallDamage);
+		this.flyingFallDamage = flyingFallDamage;
+	}
+
+	@Override
+	public @NotNull TriState hasFlyingFallDamage()
+	{
+		return this.flyingFallDamage;
+	}
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/PlayerMock.java]: These should be implemented in their own PR

File: src/main/java/be/seeseemelk/mockbukkit/entity/PoweredMinecartMock.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+import org.bukkit.Material;
+import org.bukkit.entity.minecart.PoweredMinecart;
+import org.jetbrains.annotations.NotNull;
+
+import com.google.common.base.Preconditions;
+
+import be.seeseemelk.mockbukkit.ServerMock;
+
+public class PoweredMinecartMock extends MinecartMock implements PoweredMinecart
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/PoweredMinecartMock.java:16]: Missing Javadoc

File: src/main/java/be/seeseemelk/mockbukkit/entity/PoweredMinecartMock.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+{
+
+	private double zPush;
+	private double xPush;
+	private int fuel;
+
+	public PoweredMinecartMock(@NotNull ServerMock server, @NotNull UUID uuid)
```
Comment (Insprill) [inline src/main/java/be/seeseemelk/mockbukkit/entity/PoweredMinecartMock.java:29]: Missing Javadoc

**Review by Insprill — CHANGES_REQUESTED** [review]

File: src/main/java/be/seeseemelk/mockbukkit/entity/PlayerMock.java (hunk 9 above)
Comment (Thorinwasher) [reply src/main/java/be/seeseemelk/mockbukkit/entity/PlayerMock.java]: Thought so

**Review by thelooter — APPROVED** [review]

**Review by Insprill — APPROVED** [review]
LGTM
